addons/web/tests/utils.py

```diff
@@ -0,0 +1,106 @@
+        base_url = HttpCase.base_url()
+    else:
+        base_url = f"http://{HOST}:{odoo.tools.config['http_port']}"
+    request = Mock(
+        # request
+        httprequest=Mock(
+            host='localhost',
+            path=path,
+            app=odoo.http.root,
+            environ=dict(
+                EnvironBuilder(
+                    path=path,
+                    base_url=base_url,
+                    environ_base=environ_base,
+                ).get_environ(),
+                REMOTE_ADDR=remote_addr,
+            ),
+            cookies=cookies,
+            referrer='',
+            remote_addr=remote_addr,
+            url_root=url_root,
+            args=[],
+        ),
+        type='http',
+        future_response=odoo.http.FutureResponse(),
+        params={},
+        redirect=env['ir.http']._redirect,
+        session=DotDict(
+            odoo.http.get_default_session(),
+            context={'lang': ''},
+            force_website_id=website and website.id,
+        ),
+        geoip=odoo.http.GeoIP('127.0.0.1'),
+        db=env.registry.db_name,
+        env=env,
+        registry=env.registry,
+        cr=env.cr,
+        uid=env.uid,
+        context=env.context,
+        cookies=cookies,
+        lang=env['res.lang']._get_data(code=lang_code),
+        website=website,
+        render=lambda *a, **kw: '<MockResponse>',
+    )
+    if url_root is not None:
+        request.httprequest.url = werkzeug.urls.url_join(url_root, path)
+    if website:
+        request.website_routing = website.id
+    if country_code:
+        try:
+            request.geoip._city_record = odoo.http.geoip2.models.City(['en'], country={'iso_code': country_code})
+        except TypeError:
+            request.geoip._city_record = odoo.http.geoip2.models.City({'country': {'iso_code': country_code}})
+
+    # The following code mocks match() to return a fake rule with a fake
+    # 'routing' attribute (routing=True) or to raise a NotFound
+    # exception (routing=False).
+    #
+    #   router = odoo.http.root.get_db_router()
+    #   rule, args = router.bind(...).match(path)
+    #   # arg routing is True => rule.endpoint.routing == {...}
+    #   # arg routing is False => NotFound exception
+    router = MagicMock()
+    match = router.return_value.bind.return_value.match
+    if routing:
+        match.return_value[0].routing = {
+            'type': 'http',
+            'website': True,
+            'multilang': multilang
+        }
+    else:
+        match.side_effect = NotFound
+
+    def update_context(**overrides):
+        request.env = request.env(context=dict(request.context, **overrides))
+        request.context = request.env.context
+
+    request.update_context = update_context
+
+    with contextlib.ExitStack() as s:
+        odoo.http._request_stack.push(request)
+        s.callback(odoo.http._request_stack.pop)
+        s.enter_context(patch('odoo.http.root.get_db_router', router))
+
+        yield request
```

Comment:
The naming of the file `utils.py` is triggering ne.

All tests helpers are in the `common` module, .e.g. `from odoo.tests.common import TransactionCase`, `tagged`, `users`, ...

Shouldn't this be renamed `common.py` ?

In addition, couldn't we put this simply in `odoo/tests/common.py` ?
Given `class Request` is implementend in `odoo/http.py`, it would make sense it's mocking tool would be in the core as well, along it.

Or is there a dependence on which `MockRequest` relies which is available in `web` only ?

Reply:
> In addition, couldn't we put this simply in odoo/tests/common.py ?
> Given class Request is implementend in odoo/http.py, it would make sense it's mocking tool would be in the core as well, along it.

I agree with this argument, especially as no `web` dependency is required.
I will therefore move the function.

Reply:
I will do the move in https://github.com/odoo/odoo/pull/212405 which can replace this PR. (discussed with thle)